### PR TITLE
game: Add C++ emulator example

### PR DESCRIPTION
**Pull request by BrianShTsoi**
Adds a simple program that emulates game input and output. 
This will be useful for the game C++ rewrite.

The frames used in the example are hardcoded (hence the ungodly amount of lines of code).

### CODE REVIEW

File: game/cpp_template/src/emulator.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+void print_frame(const int screen[SCREEN_HEIGHT][SCREEN_WIDTH]) {
+  // Note that the █ character takes two bytes to store, so the buffer length
+  // needs to be big
+  const int BUF_LEN = SCREEN_HEIGHT * SCREEN_WIDTH * 3;
+
+  char buf[BUF_LEN] = {0};
+
+  sprintf(buf + strlen(buf), "\x1b[?25l");
+  sprintf(buf + strlen(buf), "\x1b[H");
```
Comment (alayander) [inline game/cpp_template/src/emulator.cpp]: nit: I would define the escape sequences with constants

File: game/cpp_template/src/emulator.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    sprintf(buf + strlen(buf), "\r\n");
+  }
+
+  printf("%s", buf);
+  fflush(stdout);
+}
+
+void die(const char *s) {
+  write(STDOUT_FILENO, "\x1b[2J", 4);
+  write(STDOUT_FILENO, "\x1b[H", 3);
```
Comment (alayander) [inline game/cpp_template/src/emulator.cpp]: nit: Same as above

File: game/cpp_template/src/emulator.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    die("read");
+  }
+
+  return key;
+}
+
+void reset_terminal() {
+  if (tcsetattr(STDIN_FILENO, TCSAFLUSH, &orig_termios) == -1)
+    die("tcsetattr");
+  printf("\x1b[2J\x1b[H\x1b[?25h");
```
Comment (alayander) [inline game/cpp_template/src/emulator.cpp]: nit: Same as above

**Review by alayander — DISMISSED** [review]
I agree that template is probably not the right name here, I'd either consider adding another hierarchy into the directory structure (like `cpp/emulated_prototype` or `cpp/prototype`) or alternatively just a rename to this extent. Otherwise minor changes requested. I'll approve since everything getting checked in is just for reference anyway, so can disregard requests/comments if you don't wanna do them lol. Although I wonder if we should just check this into a separate branch on origin, since it's not functional to the main codebase (not opposed to checking into main though if you would rather do that)

**Comment by BrianShTsoi** [comment]
Changed to consts and reorganized directory structure. Also had some minor refactoring. 
I now call it emulator example, seems more appropriate since rather than serving as a starting point for the development of the game, the intent of this program is more to demonstrate how to emulate game input/output in the terminal (and offer potentially reusable code).

Making it a different branch is probably best practice, but I feel like it might be a bit too much trouble. And the same argument applies to the python game code: it's a prototype that's not functional to the final product/main codebase. So I think we can just leave them in the same branch.

**Review by alayander — APPROVED** [review]
Yeah agree with both points, for this we can just check it into main then.

LGTM

**Pull request merged**